chia/util/cached_bls.py

```diff
@@ -0,0 +1,49 @@
+import functools
+from typing import List, Optional
+
+from blspy import AugSchemeMPL, G1Element, G2Element, GTElement
+from chia.util.hash import std_hash
+from chia.util.lru_cache import LRUCache
+
+
+def get_pairings(cache: LRUCache, pks: List[G1Element], msgs: List[bytes], force: bool) -> List[GTElement]:
+    pairings: List[GTElement] = []
+    missing_count: int = 0
+    for pk, msg in zip(pks, msgs):
+        aug_msg: bytes = bytes(pk) + msg
+        h: bytes = bytes(std_hash(aug_msg))
+        pairing: Optional[GTElement] = cache.get(h)
+        if not force and pairing is None:
+            missing_count += 1
+            # Heuristic to avoid more expensive sig validation with pairing
+            # cache when it's empty and cached pairings won't be useful later
+            # (e.g. while syncing)
+            if missing_count > len(pks) // 2:
+                return []
+        pairings.append(pairing)
+
+    for i, pairing in enumerate(pairings):
+        if pairing is None:
+            aug_msg = bytes(pks[i]) + msgs[i]
+            aug_hash: G2Element = AugSchemeMPL.g2_from_message(aug_msg)
+            pairing = pks[i].pair(aug_hash)
+
+            h = bytes(std_hash(aug_msg))
+            cache.put(h, pairing)
+            pairings[i] = pairing
+
+    return pairings
+
+
+local_cache = LRUCache(10000)
+
+
+def aggregate_verify(pks: List[G1Element], msgs: List[bytes], sig: G2Element, force: bool = False):
+    if len(pks) == 0 or len(msgs) == 0:
+        return AugSchemeMPL.aggregate_verify(pks, msgs, sig)
+    pairings: List[GTElement] = get_pairings(local_cache, pks, msgs, force)
+    if len(pairings) == 0:
+        return AugSchemeMPL.aggregate_verify(pks, msgs, sig)
+
+    pairings_prod: GTElement = functools.reduce(GTElement.__mul__, pairings)
```

Comment:
But if you remove the short-circuit cases, I think you have to add `initializer=GTElement()` in case `len(pairings) == 0`.

Reply:
(And that's assuming `GTElement()` yields zero/infinity like it does for `G1Element` and `G2Element`.)

Reply:
Then we'll end up performing the expensive pairing calculation to verify the trivial case.